azurerm/resource_arm_data_factory_v2.go

```diff
@@ -0,0 +1,349 @@
+						"account_name": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"collaboration_branch": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"project_name": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"repository_name": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"root_folder": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"tenant_id": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+				ConflictsWith: []string{"github_configuration"},
+			},
+
+			"location": locationSchema(),
+
+			"resource_group_name": resourceGroupNameSchema(),
+
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func resourceArmDataFactoryV2CreateOrUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).dataFactoryClient
+	ctx := meta.(*ArmClient).StopContext
+
+	name := d.Get("name").(string)
+	location := azureRMNormalizeLocation(d.Get("location").(string))
+	resourceGroup := d.Get("resource_group_name").(string)
+	tags := d.Get("tags").(map[string]interface{})
+
+	dataFactory := datafactory.Factory{
+		Location: &location,
+		Tags:     expandTags(tags),
+	}
+
+	if v, ok := d.GetOk("identity.0.type"); ok {
+		identityType := v.(string)
+		dataFactory.Identity = &datafactory.FactoryIdentity{
+			Type: &identityType,
+		}
+	}
+
+	_, err := client.CreateOrUpdate(ctx, resourceGroup, name, dataFactory, "")
+	if err != nil {
+		return err
+	}
+
+	resp, err := client.Get(ctx, resourceGroup, name, "")
+	if err != nil {
+		return err
+	}
+
+	if resp.ID == nil {
+		return fmt.Errorf("Cannot read Data Factory %s (resource group %s) ID", name, resourceGroup)
+	}
+
+	if hasRepo, repo := expandArmDataFactoryV2RepoConfiguration(d); hasRepo {
+		repoUpdate := datafactory.FactoryRepoUpdate{
+			FactoryResourceID: resp.ID,
+			RepoConfiguration: repo,
+		}
+		_, err = client.ConfigureFactoryRepo(ctx, location, repoUpdate)
+		if err != nil {
+			return err
+		}
+	}
+
+	d.SetId(*resp.ID)
+
+	return resourceArmDataFactoryV2Read(d, meta)
+}
+
+func resourceArmDataFactoryV2Read(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).dataFactoryClient
+	ctx := meta.(*ArmClient).StopContext
+
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resourceGroup := id.ResourceGroup
+	name := id.Path["factories"]
+
+	resp, err := client.Get(ctx, resourceGroup, name, "")
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			d.SetId("")
+			return nil
+		}
+
+		return fmt.Errorf("Error making Read request on AzureRM Data Factory %s: %+v", name, err)
+	}
+
+	d.Set("name", resp.Name)
+	d.Set("resource_group_name", resourceGroup)
+	if location := resp.Location; location != nil {
+		d.Set("location", azureRMNormalizeLocation(*location))
+	}
+
+	repoType, repo := flattenArmDataFactoryV2RepoConfiguration(&resp)
```

Comment:
Could we break this out into two separate functions? one for github and one for vsts

Reply:
Any particular reason? It's kind of tricky as you have to drill down into the properties before you even find out what repo type you are dealing with.
Happy to make the change if there's a better way to do it